### PR TITLE
lib: posix: nanosleep

### LINKED ISSUE #25554: lib: posix: nanosleep
**Is your enhancement proposal related to a problem? Please describe.**
nanosleep(2) is required for std::chrono in C++, among other things. It's POSIX (IEEE 1003.1) as well.

**Describe the solution you'd like**
I'll create a PR.

**Describe alternatives you've considered**
Looked into newlib. There is no de-facto implementation of nanosleep. It generally just makes a system call. FWICT, Zephyr only has k_msleep() precision, which relies on ticks, FWICT.

**Additional context**
Loosely related to #24730 (where reusing time.h header from libc was mentioned)
Required by #25527

### PR DESCRIPTION
**Pull request by cfriedt**
This change adds support for nanosleep(2) documented at
https://pubs.opengroup.org/onlinepubs/9699919799/functions/nanosleep.html

Note: Does not achieve any greater resolution than that provided by `k_busy_wait(us)` but at least makes the API available.

I've not researched it much in Zephyr, but I would say there is still some work to be done for a generic clock API - one that allows querying resolutions, supports multiple system clocks, etc, primarily to support POSIX time APIs.

A nanosleep system call would also require high-res timer support (See #6498)

Fixes #25554

### CODE REVIEW

File: lib/posix/sleep.c
Comment (pfalcon) [inline lib/posix/sleep.c:18]: It's good zen exercise to try to avoid patching unrelated (== not directly related) code ;-).

Comment (cfriedt) [reply lib/posix/sleep.c:18]: Will fix. That was a checkpatch.pl / fix-inline modification.

**Comment by cfriedt** [comment]
I have a feeling there is a conflict with `CONFIG_BOARD_NATIVE_POSIX`
`CONFIG_BOARD_NATIVE_POSIX_64BIT`

**Comment by pfalcon** [comment]
> I have a feeling there is a conflict with CONFIG_BOARD_NATIVE_POSIX
CONFIG_BOARD_NATIVE_POSIX_64BIT

I'm not sure what exactly you mean (that native_posix and native_posix_64 boards conflict between each other?), but otherwise it's a known problem that Zephyr POSIX subsys conflicts with native_posix board. It should be understanding why - it's too layery a pie, where Zephyr POSIX subsys sits on top of native Zephyr core, and naive_posix boards tries to implement that in terms of the underlying full-POSIX system. Something doesn't match regularly. Many a POSIX tests/samples simply blacklist native_posix.

**Comment by cfriedt** [comment]
> > I have a feeling there is a conflict with CONFIG_BOARD_NATIVE_POSIX
> > CONFIG_BOARD_NATIVE_POSIX_64BIT
> 
> I'm not sure what exactly you mean (that native_posix and native_posix_64 boards conflict between each other?), but otherwise it's a known problem that Zephyr POSIX subsys conflicts with native_posix board. It should be understanding why - it's too layery a pie, where Zephyr POSIX subsys sits on top 

That's what I meant. Sorry, was on mobile when I commented previously.

**Comment by cfriedt** [comment]
Should be good to go now
```
$ ./scripts/sanitycheck --enable-slow --inline-logs -j8 --testcase-root tests/posix/common
Renaming output directory to /home/cfriedt/workspace/zephyrproject/zephyr/sanity-out.24
INFO    - JOBS: 8
INFO    - Selecting default platforms per test case
INFO    - Building initial testcase list...
INFO    - 26 test configurations selected, 972 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
INFO    - Total complete:   26/  26  100%  skipped:    0, failed:    0
INFO    - 26 of 26 tests passed (100.00%), 0 failed, 0 skipped with 0 warnings in 179.42 seconds
INFO    - In total 4 test cases were executed on 13 out of total 250 platforms (5.20%)
INFO    - 20 tests executed on platforms, 6 tests were only built.
```

**Comment by cfriedt** [comment]
Btw, the variance in latency (in CI / qemu) is pretty drastic between architectures. E.g. on qemu_cortex_r5, it was 500us, on xtensa it was 86 us, and on qemu_cortex_m3 it was 25us.

Sample size of 1, not statistically representative, of course.

File: tests/posix/common/src/nanosleep.c
Comment (cfriedt) [inline tests/posix/common/src/nanosleep.c:190]: Looks like I need the ceil variant here. Although ns (realistically??)

File: lib/posix/nanosleep.c
Comment (pfalcon) [inline lib/posix/nanosleep.c:37]: Label name sounds surprising. It's rather "return_val" or "skip" to me. Probably not worth changing, just thinking aloud.

**Review by pfalcon — APPROVED** [review]
Thanks!

**Comment by pfalcon** [comment]
@cfriedt, Thanks for the patch. I would personally recommend this merged for 2.3, but per the release freeze rules, 2 reviews are required at this time. I've added @carlescufi who both could review and merged (but definitely busy with a lot of other stuff now, as the release manager for this version), but feel free ping other folks who you know can help with review.

**Comment by cfriedt** [comment]
> @cfriedt, Thanks for the patch. I would personally recommend this merged for 2.3, but per the release freeze rules, 2 reviews are required at this time. I've added @carlescufi who both could review and merged (but definitely busy with a lot of other stuff now, as the release manager for this version), but feel free ping other folks who you know can help with review.

It would be awesome if nanosleep could actually get relatively close to the exact time (if there is hardware to support it). The high-res timer stuff is pretty important though, and I think it deserves a great deal of care. No rush on that one. The nice thing about it though is that the interface is there, and it's not moving, so just redoing the implementation later will have no unwanted side-effects.

File: lib/posix/nanosleep.c
Comment (jukkar) [inline lib/posix/nanosleep.c:13]: There is `NSEC_PER_SEC` defined in `include/sys_clock.h` so we should use that

File: lib/posix/nanosleep.c
Comment (jukkar) [inline lib/posix/nanosleep.c:14]: Also `NSEC_PER_USEC` is also defined so lets use that instead

File: lib/posix/nanosleep.c
Comment (jukkar) [inline lib/posix/nanosleep.c:24]: Better would be something like `do_write` so that when grepping for `write` function we would not get this one.
Also () are superfluous and could be removed, but if you think it improves readability then I am fine with this too.

File: lib/posix/nanosleep.c
Comment (jukkar) [inline lib/posix/nanosleep.c:29]: please place empty lines after } ending a block

File: lib/posix/nanosleep.c
Comment (jukkar) [inline lib/posix/nanosleep.c:51]: Hmm, could we still call `k_sleep()` as it supports us sleep values, not sure if it can actually sleep that short time.

**Review by jukkar — CHANGES_REQUESTED** [review]

File: lib/posix/nanosleep.c
Comment (pfalcon) [reply lib/posix/nanosleep.c:51]: I assume the usecase for nanosleep() is really short intervals, which also should be as precise as possible, that's why @cfriedt used k_busy_wait(). k_sleep() leads to rescheduling of the thread, so any hopes for precision are lost.

File: tests/posix/common/src/nanosleep.c
Comment (jukkar) [inline tests/posix/common/src/nanosleep.c:11]: We should use here `NSEC_PER_SEC`

**Review by jukkar — APPROVED** [review]
Looks good, minor nit about the macro name that was missed from previous cleanup.

File: tests/posix/common/src/nanosleep.c
Comment (cfriedt) [reply tests/posix/common/src/nanosleep.c:11]: Sorry, I thought I had fixed that already, thanks for pointing it out.

**Comment by cfriedt** [comment]
@carlescufi  - rebased off of using `<stdint.h>` integer types. Should be good to go as long as the build works.

**Comment by cfriedt** [comment]
Was green before. Looks like there is some misra change that breaks things atm

**Comment by pfalcon** [comment]
> some misra change

New swear word in software engineering? ;-)

**Comment by pfalcon** [comment]
~~~
starting test - test_timeout_lifo_thread

    Assertion failed at WEST_TOPDIR/zephyr/tests/kernel/lifo/lifo_usage/src/main.c:329: test_timeout_lifo_thread: packet != NULL is false

FAIL - test_timeout_lifo_thread
~~~

I got such a failure in one of my recent PR. Looks like nondeterministic failure du jour.

@cfriedt: Please keep rebasing this, until sweet spot is found.

**Pull request merged**
